Blog: Publishing Modbus data to your UNS with FlowFuse (#2788)
FlowFuse/website: src/blog/2024/11/publishing-modbus-data-to-uns.md

   - modbus
   - uns
---

As industries evolve towards Industry 4.0, integrating legacy systems like Modbus with modern cloud technologies becomes essential. While effective for local machine-to-machine communication, Modbus doesn't naturally connect with cloud platforms.

This article shows you how to bridge Modbus data to MQTT using FlowFuse and Node-RED. By publishing this data into a [Unified Namespace (UNS)](/solutions/uns/), you can seamlessly integrate legacy systems with cloud services for real-time monitoring and analysis, unlocking the full potential of your industrial data.

<!--more-->

## The Importance of Bridging Modbus Data to MQTT for Cloud Integration and UNS

Modbus has long been a foundational protocol in industrial environments, enabling reliable communication between devices like PLCs, sensors, and motors. It has served the needs of operational technology (OT) for decades. However, as industries embrace Industry 4.0, there is an increasing need to bridge the gap between legacy OT systems like Modbus and modern information technology (IT) platforms, including cloud-based systems and IoT applications. This is where the challenge of the OT/IT gap arises.

> While Modbus excels in local control and data collection, it struggles to integrate with cloud-based platforms and real-time IoT solutions. That's where MQTT, a lightweight, real-time messaging protocol, becomes essential. MQTT allows devices to transmit data efficiently to the cloud, even in environments with limited bandwidth, power, or connectivity. However, bridging them is crucial since Modbus and MQTT don’t natively communicate for bridging. We will need to use a platform like Flowfuse that supports both.

Reviewer: This paragraph needs to be about UNS, not MQTT. MQTT is a broker _choice_ for the UNS which is the data architecture to communicate with arbitrary consumers. For our UNS we use MQTT, which then links to your other post about setting up a UNS in 15 minutes?

Connecting Modbus data to MQTT allows you to send data from legacy systems to the cloud without replacing existing equipment. This enables real-time monitoring and remote analysis. Once you create the bridge for data from Modbus to MQTT, publishing it into a Unified Namespace (UNS) further enhances data accessibility and usability. A UNS centralizes all operational data into a single namespace, making it easier to track performance, detect issues, and perform analytics without complex integrations.

For instance, in a factory still relying on Modbus, machines such as motors, pumps, and conveyors can send their operational data to the cloud via MQTT. Once the data is in the cloud, it can be published into the Unified Namespace (UNS), giving stakeholders easy access to a centralized view of all device data in real time. This integration provides immediate insights into machine performance, enabling operators to track temperature, vibration, and runtime metrics. With these insights, predictive maintenance models can be applied, helping to identify potential failures before they occur and preventing unplanned downtime.

## How to Bridge Modbus to MQTT

> This section will cover the steps to bridge Modbus data to MQTT using FlowFuse, leveraging Node-RED's capabilities. The process involves retrieving data from a Modbus device (For Practical example, we are using OpenSim to simulate Modbus data), transforming and processing the data (e.g., scaling raw sensor data into human-readable formats), and sending it to your Unified Namespace.

Reviewer 2: ```suggestion
Let's look at the steps to bridge Modbus data to MQTT using FlowFuse, leveraging Node-RED's capabilities. The process involves retrieving data from a Modbus device (For Practical example, we are using OpenSim to simulate Modbus data), transforming and processing the data (e.g., scaling raw sensor data into human-readable formats), and sending it to your Unified Namespace.
```
